tests/test_session.py

```diff
@@ -88,7 +87,9 @@ def test_reconnect_and_commit(initial_operations):
 def test_get_cached_collections(initial_operations):
     SESSION.reconnect_and_commit(initial_operations[0])
 
-    users = SESSION.get_cached_collections()[User].fetch_attributes("first_name", "last_name", "is_active")
-
-    users_split = users.partition(lambda u: u.username[0] == initial_operations[1]["username"])
-    assert users_split[0].first_name == [initial_operations[1]["first_name"]]
+    cached_users = SESSION.get_cached_collections()[User]
+    for cached_user in cached_users:
+        user = Query(User).by_id(cached_user.id[0]).get_first(projections=["username", "first_name"])
```

Comment:
Isn't the purpose of the cache that we won't have to query again?

Reply:
I did not write the test but I guess the query is to make sure the user actually exists and comparing it with the user from "initial_operations". If you compare my changes to what it was before, I did not actually change anything in the test itself, just added an additional check for the user's existence.

Reply:
Do you mean you took over the existing changes from @salvaom?

I'm still wondering why the Query call has been added because for me it defeats the purpose of this test.

Reply:
the query is added as an additional check for the user's existence. _get_cached_collections_ returns two user ids, the first one of which does not exist and will cause the test to fail.

Reply:
Still, the user shouldn't be taken from the query result but from the cache when it exists.

Reply:
What are you referring to with "when it exists" - the user or the cache?

this line causes ftrack to loop endlessly and fail because get_cached_collections returns a nonexisting user id:
https://github.com/TrixterFilm/trackteroid/blob/5467820e458947c62118bd6ed8cbebe5f1dc5318/tests/test_session.py#L91C1-L92C1

Since I am not familiar with get_cached_collections, is that expected behaviour? -> If yes, what other verification method do you suggest to verify the existence of the user other than a query for its ID?
-> if no, the test should be made to fail at that point with a proper assert/without ftrack crashing and we should open a ticket to fix "get_cached_collections"

please advise

Reply:
thank you.

> Just to be sure here: `get_cached_collections` returns a collection, not an id/ids, correct?

that is correct. During the test, it returns a collection of two users, the first of which does not exist, the other being the correct one. I have not investigated further as to what that first user is - so the question remains if that is expected behaviour or not.

> is it `fetch_attributes` that causes an endless loop?

Exactly, therefore I separated it out - so `Query...get_first()` will return an `EmptyCollection` if a user does not exist and `fetch_attributes` is only called with the second valid user.

Reply:
>  During the test, it returns a collection of two users, the first of which does not exist, the other being the correct one. I have not investigated further as to what that first user is - so the question remains if that is expected behaviour or not.

Looking into the fixture `initial_operations` I think it makes sense what is happening, although I wouldn't consider this "expected" behavior when a deletion on the server happened. But as we use the `SESSION` singleton across the test functions,  this singleton still contains a reference to a user that was deleted within a prior test function call.

I think what needs to happen is:

1. Change the tests so they won't use the `SESSION` singleton
2. Identify why [self._local_cache](https://github.com/TrixterFilm/trackteroid/blob/main/src/trackteroid/session.py#L135C91-L135C117) isn't properly updated when a deletion operation was successfully committed
3. Also, identify why `fetch_attributes` is causing an endless loop, this should definitely not happen.

Once you confirm that my assumptions here are correct, I'd suggest tracking 2. and 3. via individual new tickets.

Reply:
>1. Change the tests so they won't use the SESSION singleton

that will not be possible as it is the `SESSION`'s functions the tests are testing and depend upon - `reconnect_and_commit` for example. If the behaviour is not expected, the proper way is to expose it in the test so it is failing properly and not crashing in ftrack should it happen again. I will add that.

For 2 and 3, I will add the tickets

Reply:
> that will not be possible as it is the SESSION's functions the tests are testing and depend upon - reconnect_and_commit for example

I'm just referring to not using the singleton session instance, so the tests should look like this:

```
    session = Session()
    session.reconnect_and_commit(initial_operations[0])

    cached_users = session.get_cached_collections()[User]
```

Currently, the entity is being cleaned up [using a new Ftrack session instance](https://github.com/TrixterFilm/trackteroid/blob/main/tests/test_session.py#L43) which explains why the local cache of the `SESSION` doesn't reflect the deletion.

Reply:
Can you maybe point me to your changes?

Reply:
The problem is that you don't pass the session object to the query. Remember that Query is taking the `SESSION` singleton instance by default.

Let me know if this solves the issue.

Reply:
thank you, that solves the problem for  test_deferred_operations.

Still getting that "mystery user" - so I just saw that there are dbm files checked in under tests/resources/session - which when deleted makes all the tests fail (and which probably contain that nonexisting user). 
This dbm file should be created by the test and deleted thereafter for the test to be valid, which I think was done in the original test_session.py before trackteroid.

In order to proceed, I need to understand how the original tests were working. Since it is not quite obvious in the code, I woud need some clarifications here:

```
import anydbm
import os
import logging
import tempfile
import shutil
import uuid

from trixter.ftrack import (
    Query,
    SESSION,
)

from trixter.ftrack.session import (
    CreateEntityOperation,
    UpdateEntityOperation
)

from trixter.ftrack.entities import *

from .base import TestCase

_LOG = logging.getLogger("tx.test.ftrack.session")


class TestSessionDelegate(TestCase):
    auto_populate = False
    server_url = "http://ftrack-devel.trixter.intern"

    @classmethod
    def setUpClass(cls):
        super(TestSessionDelegate, cls).setUpClass()
        cls.tmp_dir = os.path.join(tempfile.gettempdir(), str(uuid.uuid4()))

        os.makedirs(cls.tmp_dir)

        cls.dumped_operations_file = os.path.join(cls.tmp_dir, "operations.dbm")
        cls.initial_operations_file = os.path.join(os.path.dirname(__file__), "fixtures", "operations.dbm")

    @classmethod
    def tearDownClass(cls):
        shutil.rmtree(cls.tmp_dir)

    @classmethod
    def setUp(cls):
        Query(AssetVersion).by_id("fbb682b6-e9e6-4111-8edb-38d0797c9ffe").get_first().delete().commit()
```
Where is that assetversion coming from? Why is it deleted on "setup"


```
    def test_deferred_operations(self):

        # case 1. clear an existing cache temporarily
        query_result = self.samplebag.grab("AssetVersion", attributes=["task", "version"])
        operations_count = len(SESSION.recorded_operations)

        # do a query -> cache a result
        with SESSION.deferred_operations(self.dumped_operations_file):
            self.assertEqual(0, len(SESSION.recorded_operations))

            avs = Query(AssetVersion).by_id(query_result["id"]).get_one(projections=["task"])
            avs.version = avs.version[0] + 10
            avs2 = avs.create(task=avs.task)  # -> create entity, update task, update asset
            avs2.version = avs.version[0] + 10
            
            self.assertEqual(5, len(SESSION.recorded_operations))

        self.assertEqual(operations_count, len(SESSION.recorded_operations))
        # check the created file database
        database = anydbm.open(self.dumped_operations_file, "r")
        def make_keys(entity_collection):
            return entity_collection.map(
                lambda x: "('{}', ['{}'])".format(x.entity_type.__name__, x.id[0])
            )

        expected_keys = make_keys(avs.union(avs2)) + make_keys(avs.task) + make_keys(avs.asset) + ["__operations__"]
        self.assertItemsEqual(expected_keys, database.keys())

    def test_reconnect_and_commit(self):

        SESSION.reconnect_and_commit(
            self.initial_operations_file,
            server_url=self.server_url
        )

        assetversion = Query(AssetVersion).by_id("fbb682b6-e9e6-4111-8edb-38d0797c9ffe").get_one(
            projections=["components.name"]
        )
        self.assertEqual(99, assetversion.version[0])
        self.assertEqual("pymelle", assetversion.components.name[0])
```
Again, where/when is that Assetversion/Component created?

```
    def test_get_cached_collections(self):

        SESSION.reconnect_and_commit(
            self.initial_operations_file,
            server_url=self.server_url
        )

        avs = SESSION.get_cached_collections()[AssetVersion]
        avs.fetch_attributes("asset.versions")

        v99, rest = avs.asset.versions.partition(lambda x: x.version[0] == 99)
        v99.uses_versions = rest

        SESSION.commit()
        ```

Also, I do not see code that writes the dbm file with certain operations. Where/how does that happen?

Reply:
> Still getting that "mystery user" - so I just saw that there are dbm files checked in under tests/resources/session - which when deleted makes all the tests fail (and which probably contain that nonexisting user).

Yes, the dbm file is the fixture/dataset that has to correlate with expectations in the test code. So you shouldn't delete them and either stick to the operations this database contains (the user creation) or regenerate other dbm files with your relevant changes that correlate with your adjustments. Please also note that the dbm files are different for linux and windows, which is a requirement for cross-platform testing.

Be aware that it is not trivial to change the dataset as many operations rely on parent ids that are not reproducible across test runs. This was one problem I stumbled on and found the User to be an entity which serves our test purpose quite well.

> This dbm file should be created by the test and deleted thereafter for the test to be valid, which I think was done in the original test_session.py before trackteroid.

What tests are you referring to exactly? From what I can see in the code it is how it should be:

We have [_initial operations_](https://github.com/TrixterFilm/trackteroid/blob/a09bc1e66bff54cdc37001bf648a7afb37d58c78/tests/test_session.py#L78) which is a pre-dumped file that contains the operations we test against.

And then we have [_deferred operations_](https://github.com/TrixterFilm/trackteroid/blob/a09bc1e66bff54cdc37001bf648a7afb37d58c78/tests/test_session.py#L47) that is a temporary file path and these database files will be created by some tests itself.

Pretty identical to the original test cases you've pasted:
```
        cls.dumped_operations_file = os.path.join(cls.tmp_dir, "operations.dbm")
        cls.initial_operations_file = os.path.join(os.path.dirname(__file__), "fixtures", "operations.dbm")
```

Reply:
cool, so the dbm file needs to be pre-created and is not created as part of the tests, that clarifies it, thank you.

Reply:
> cool, so the dbm file needs to be pre-created and is not created as part of the tests, that clarifies it, thank you.

Yes, for what we call _initial operations_.